crates/turborepo-lockfiles/src/bun/mod.rs

```diff
@@ -0,0 +1,173 @@
+use std::str::FromStr;
+
+use serde::Deserialize;
+
+use crate::Lockfile;
+
+mod de;
+mod ser;
+
+type Map<K, V> = std::collections::BTreeMap<K, V>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("unable to parse: {0}")]
+    SymlParse(String),
+    #[error("unable to convert to structured syml: {0}")]
+    SymlStructure(#[from] serde_json::Error),
+    #[error("unexpected non-utf8 yarn.lock")]
+    NonUTF8(#[from] std::str::Utf8Error),
+}
+
+pub struct BunLockfile {
+    inner: Map<String, Entry>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Entry {
+    name: Option<String>,
+    version: String,
+    uid: Option<String>,
+    resolved: Option<String>,
+    integrity: Option<String>,
+    registry: Option<String>,
+    dependencies: Option<Map<String, String>>,
+    optional_dependencies: Option<Map<String, String>>,
+}
+
+impl BunLockfile {
+    pub fn from_bytes(input: &[u8]) -> Result<Self, super::Error> {
+        let input = std::str::from_utf8(input).map_err(Error::from)?;
+        Self::from_str(input)
+    }
+}
+
+impl FromStr for BunLockfile {
+    type Err = super::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let value = de::parse_syml(s)?;
+        let inner = serde_json::from_value(value)?;
+        Ok(Self { inner })
+    }
+}
+
+impl Lockfile for BunLockfile {
+    fn resolve_package(
+        &self,
+        _workspace_path: &str,
+        name: &str,
+        version: &str,
+    ) -> Result<Option<crate::Package>, crate::Error> {
+        for key in possible_keys(name, version) {
+            if let Some(entry) = self.inner.get(&key) {
+                return Ok(Some(crate::Package {
+                    key,
+                    version: entry.version.clone(),
+                }));
+            }
+        }
+
+        Ok(None)
+    }
+
+    fn all_dependencies(
+        &self,
+        key: &str,
+    ) -> Result<Option<std::collections::HashMap<String, String>>, crate::Error> {
+        let Some(entry) = self.inner.get(key) else {
+            return Ok(None);
+        };
+
+        let all_deps: std::collections::HashMap<_, _> = entry.dependency_entries().collect();
+        Ok(match all_deps.is_empty() {
+            false => Some(all_deps),
+            true => None,
+        })
+    }
+
+    fn subgraph(
+        &self,
+        _workspace_packages: &[String],
+        packages: &[String],
+    ) -> Result<Box<dyn Lockfile>, super::Error> {
+        let mut inner = Map::new();
+
+        for (key, entry) in packages.iter().filter_map(|key| {
+            let entry = self.inner.get(key)?;
+            Some((key, entry))
+        }) {
+            inner.insert(key.clone(), entry.clone());
+        }
+
+        Ok(Box::new(Self { inner }))
+    }
+
+    fn encode(&self) -> Result<Vec<u8>, crate::Error> {
+        Ok(self.to_string().into_bytes())
```

Comment:
Yeah, I was basically intending to be cute and hand users back a pruned `yarn` project if they started with a `bun` one. I don't know when we're going to get to `ser` but since it works to prune to `yarn` I was trying to figure out the right tradeoff.

1. Silently prune to `yarn1`. (Current state.)
2. Prune with a large `WARNING: You must use yarn in the pruned project`
3. Disable `prune` until we get to `ser` with a fiddly workaround which is basically "lie to `turbo`" and pretend to be `yarn` to get a pruned `yarn` project.

As to `prune.rs`: 👍 but I was going to just call it "working" and therefore this copy/paste would be fine.

Reply:
(I turned `prune` off.)